pool/swap.gno

```diff
@@ -0,0 +1,513 @@
+package pool
+
+import (
+	"std"
+
+	"gno.land/p/demo/ufmt"
+
+	"gno.land/r/gnoswap/v1/common"
+	"gno.land/r/gnoswap/v1/consts"
+
+	plp "gno.land/p/gnoswap/pool" // pool package
+
+	i256 "gno.land/p/gnoswap/int256"
+	u256 "gno.land/p/gnoswap/uint256"
+)
+
+// Swap swaps token0 for token1, or token1 for token0
+// Returns swapped amount0, amount1 in string
+// ref: https://docs.gnoswap.io/contracts/pool/pool.gno#swap
+func Swap(
+	token0Path string,
+	token1Path string,
+	fee uint32,
+	recipient std.Address,
+	zeroForOne bool,
+	amountSpecified string,
+	sqrtPriceLimitX96 string,
+	payer std.Address, // router
+) (string, string) {
+	common.IsHalted()
+	if common.GetLimitCaller() {
+		caller := std.PrevRealm().Addr()
+		if err := common.RouterOnly(caller); err != nil {
+			panic(addDetailToError(
+				errNoPermission,
+				ufmt.Sprintf("pool.gno__Swap() || only router(%s) can call pool swap(), called from %s", consts.ROUTER_ADDR, caller.String()),
+			))
+		}
+	}
+
+	if amountSpecified == "0" {
+		panic(addDetailToError(
+			errInvalidSwapAmount,
+			ufmt.Sprintf("pool.gno__Swap() || amountSpecified == 0"),
+		))
+	}
+
+	pool := GetPool(token0Path, token1Path, fee)
+
+	slot0Start := pool.slot0
+	if !slot0Start.unlocked {
+		panic(errLockedPool)
+	}
+
+	slot0Start.unlocked = false
```

Comment:
@notJoon 
This syntax requires slot0Start.unlocked to be unlocked, not false, and pool.slot0 to be unlocked.
This syntax is the logic that causes pool.slot0 to copy the slot0 value of the singleton-like instance of slot0 to slot0Start and use it for computation. 
Therefore, the value of unlocked in pool.slot0 should be false, not slo0Start, so that when a re-entry attack comes in, it can be prevented by the panic code above, so it needs to be changed.

Reply:
defer func should be also modified